### PR TITLE
Introducing new current filesystem

### PR DESCRIPTION
**Pull request by pdler**
This PR is the current final state of the refactoring of the saros resources. In issue https://github.com/saros-project/saros/issues/177, I described how the new saros resources will work and I listed all coming PR and described them a little. In the following PRs, I will summarize, which commits are only refactorings or replacing of references, or in which commits I changed some logic stuff. In every PR, I will set a reference to this PR, so you can see, how the final state looks like and may better unterstand the steps, which I did. 

The implementation of the new saros resources in Saros Server is not done yet. I will do this next

### CODE REVIEW

File: de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/ISarosSession.java [hunk 1]
```diff
@@ -48,11 +48,11 @@
...
    * <p>This Interface is the main entrance Point for the "Session Management"-Component. The
-   * Session Management is responsible for managing a Session and keeping the shared projects in a
-   * consistent state across the local copies of all participants. It functions as the core
-   * component in a running session and directs communication between all other components. In
-   * general this component takes input from the User Interface, processes it, and afterwards passes
-   * the result to the Network Layer.
+   * Session Management is responsible for managing a Session and keeping the shared reference point
+   * in a consistent state across the local * copies of all participants. It functions as the core *
+   * component in a running session and directs communication * between all other components. In
+   * general this component * takes input from the User Interface, processes it, and * afterwards
```
[the remark this answers is not in the document]
Comment (pdler) [reply de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/ISarosSession.java]: It seems that the formatting is corrupt in this commit. I will fix it

**Comment by chrismotian** [comment]
I would appreciate it if you can provide a list of files that are only refactorings or replacing of references and another list with files where you changed some logic stuff. Reviewing the commits isn't a good way in my opinion because I don't know if the file will change in one of the future commits that already in this PR. Reviewing the files in the most recent version on the other hand (without the list) is a hard one because I can't decide how much I have to concentrate.

Another idea: every reviewer can take some files, review it and afterward write a comment for the other reviewers to beware of duplicate work.  I will start: I checked ISarosSession and ISarosSessionManager.

**Comment by tobous** [comment]
@pdler Can you please fix the PR so that the build succeeds? Beforehand, it doesn't really make sense to review it.

**Comment by pdler** [comment]
@tobous I'm sorry, that it took so much, but ich fixed the formatting issues and implemented the new saros core resource interface for Saros Server.

**Comment by pdler** [comment]
@chrismotian  I created a list in https://github.com/saros-project/saros/issues/177 which classes a completely new or in which class are only (automatic) refactorings, or etc..

**Comment by chrismotian** [comment]
I've checked:
/de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/filesystem/IFolder.java
/de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/filesystem/IReferencePoint.java
/de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/filesystem/ReferencePointImpl.java
/de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/IReferencePointManager.java
/de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/ReferencePointManager.java
/de.fu_berlin.inf.dpp/src/de/fu_berlin/inf/dpp/filesystem/EclipseAbstractFolderImpl.java
/de.fu_berlin.inf.dpp/src/de/fu_berlin/inf/dpp/filesystem/EclipseFolderImpl.java
/de.fu_berlin.inf.dpp/src/de/fu_berlin/inf/dpp/filesystem/EclipseProjectImpl.java

File: de.fu_berlin.inf.dpp/src/de/fu_berlin/inf/dpp/filesystem/EclipseReferencePointManager.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+   * @param referencePoint The reference point, on which the resource belongs to
+   * @param referencePointRelativePath the relative path from the reference point to the resource
+   * @return the resource of the reference point from referencePointRelativePath or null if resource
+   *     does not exist.
+   */
+  public synchronized IResource getResource(
+      IReferencePoint referencePoint, IPath referencePointRelativePath) {
+    if (referencePoint == null) throw new NullPointerException("Reference point is null");
+
+    if (referencePointRelativePath == null) throw new NullPointerException("Path is null");
```
Comment (chrismotian) [inline de.fu_berlin.inf.dpp/src/de/fu_berlin/inf/dpp/filesystem/EclipseReferencePointManager.java]: I would prefer IllegalArgumentException

**Comment by pdler** [comment]
> Can you tell me the reason for changing the types? For example: IResource.FILE to org.eclipse.core.resources.IResource.FILE
> or IResource.FOLDER to org.eclipse.core.resources.IResource.FOLDER

Because the resource has the type org.eclipse.core.resources. So I was wondering why the type of the resource was compared with constant values of saros core resource types. This only worked before because the constant values in saros core resources were the same as the eclipse resources. 
But with the new saros resources, the comparison don't work anymore, because Saros does'nt know project anymore. So the type of the eclipse resource in ResourceAdapterFactory has to be compared with the eclipse resource constant values now.

File: de.fu_berlin.inf.dpp/src/de/fu_berlin/inf/dpp/filesystem/EclipseReferencePointManager.java (hunk 2 above)
Comment (pdler) [reply de.fu_berlin.inf.dpp/src/de/fu_berlin/inf/dpp/filesystem/EclipseReferencePointManager.java]: Yes, this make more sense...Fixed!

File: de.fu_berlin.inf.dpp/src/de/fu_berlin/inf/dpp/filesystem/EclipseAbstractFolderImpl.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
+package de.fu_berlin.inf.dpp.filesystem;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import org.eclipse.core.runtime.CoreException;
+import sun.reflect.generics.reflectiveObjects.NotImplementedException;
```
[the remark this answers is not in the document]
Comment (pdler) [reply de.fu_berlin.inf.dpp/src/de/fu_berlin/inf/dpp/filesystem/EclipseAbstractFolderImpl.java]: ok, I changed to org.apache.commons.lang3.NotImplementedException

File: de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/internal/SarosSession.java [hunk 4]
```diff
@@ -1053,15 +1092,18 @@ private SarosSession(
...
 
     context.initComponent(this);
 
     this.sessionID = id;
-    this.projectMapper = new SharedProjectMapper();
+    this.sharedReferencePointMapper = new SharedReferencePointMapper();
     this.activityQueuer = new ActivityQueuer();
     this.containerContext = context;
 
+    if (referencePointManager == null) referencePointManager = new ReferencePointManager();
```
[the remark this answers is not in the document]
Comment (pdler) [reply de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/internal/SarosSession.java]: I removed this call and made sure that the referencepointmanager is never null

File: de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/internal/SarosSession.java [hunk 5]
```diff
@@ -1091,7 +1133,7 @@ private SarosSession(
     sessionContainer = context.createChildContainer();
     sessionContainer.addComponent(ISarosSession.class, this);
     sessionContainer.addComponent(IActivityHandlerCallback.class, activityCallback);
-
+    sessionContainer.addComponent(IReferencePointManager.class, new ReferencePointManager());
```
[the remark this answers is not in the document]
Comment (pdler) [reply de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/internal/SarosSession.java]: You're right. I replaced the new ReferencePointManager instance by the CTOR parameter

File: de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/ReferencePointManager.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+import java.util.HashSet;
+import java.util.Set;
+
+/** The implementation of {@link IReferencePointManager} */
+public class ReferencePointManager implements IReferencePointManager {
+
+  HashMap<IReferencePoint, IFolder> referencePointToProjectMapper;
+
+  public ReferencePointManager() {
+    referencePointToProjectMapper = new HashMap<IReferencePoint, IFolder>();
```
[the remark this answers is not in the document]
Comment (pdler) [reply de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/ReferencePointManager.java]: Is this a good practice? If you want it, I can move the initialization...

File: de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/ReferencePointManager.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+import de.fu_berlin.inf.dpp.filesystem.IFolder;
+import de.fu_berlin.inf.dpp.filesystem.IReferencePoint;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Set;
+
+/** The implementation of {@link IReferencePointManager} */
+public class ReferencePointManager implements IReferencePointManager {
+
+  HashMap<IReferencePoint, IFolder> referencePointToProjectMapper;
```
[the remark this answers is not in the document]
Comment (pdler) [reply de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/ReferencePointManager.java]: done... 
I didnt thought about using concurrent hashmap here. This is a good idea and adjusted all referencePointManager..

File: de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/ReferencePointManager.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+  public ReferencePointManager() {
+    referencePointToProjectMapper = new HashMap<IReferencePoint, IFolder>();
+  }
+
+  @Override
+  public synchronized void put(IReferencePoint referencePoint, IFolder project) {
+    if (!referencePointToProjectMapper.containsKey(referencePoint)) {
```
[the remark this answers is not in the document]
Comment (pdler) [reply de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/ReferencePointManager.java]: done

File: de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/SarosSessionManager.java [hunk 9]
```diff
@@ -141,7 +120,11 @@ public void negotiationTerminated(final ProjectNegotiation negotiation) {
           }
         }
       };
-
+  private Thread nextProjectNegotiationWorker;
```
[the remark this answers is not in the document]
Comment (pdler) [reply de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/SarosSessionManager.java]: I am sorry for this..

File: de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/internal/SharedReferencePointMapper.java [hunk 10]
```diff
@@ -0,0 +1,470 @@
...
+ * <p>The reference point IDs are used to identify shared reference point across the network, even
+ * when the local names of shared reference point are different. The ID is determined by the
+ * reference point/file-host.
+ */
+class SharedReferencePointMapper {
+
+  private static final Logger LOG = Logger.getLogger(SharedReferencePointMapper.class);
+
+  /** Mapping from reference point IDs to currently registered shared reference point. */
+  private Map<String, IReferencePoint> idToReferencePointMapping;
```
[the remark this answers is not in the document]
Comment (pdler) [reply de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/internal/SharedReferencePointMapper.java]: done

File: de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/core/ui/util/CollaborationUtils.java [hunk 11]
```diff
@@ -282,135 +348,177 @@ private static String getShareProjectDescription(ISarosSession sarosSession) {
...
+   * module {@link Module module} to {@link IFolder saros folder}
    *
-   * @param selectedResources
+   * @param selectedModules
    * @param sarosSession
    * @return
    */
-  private static Map<IProject, List<IResource>> acquireResources(
-      List<IResource> selectedResources, ISarosSession sarosSession) {
+  private static Map<IFolder, List<IResource>> acquireResourcesFromModules(
```
[the remark this answers is not in the document]
Comment (pdler) [reply de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/core/ui/util/CollaborationUtils.java]: fixed

**Comment by m273d15** [comment]
Is there a reason to keep this PR open? As far as I know #466 is the PR that has to be reviewed.

**Pull request closed without merging**